exercises/src/main/scala/fpinscala/state/State.scala

```diff
@@ -30,34 +30,118 @@ object RNG {
       (f(a), rng2)
     }
 
-  def nonNegativeInt(rng: RNG): (Int, RNG) = ???
+  // 6.1
+  // Use RNG.nextInt to generate a random number between 0 and Int.MaxValue inclusive.  Corner case
+  // when nextInt returns Int.MinValue, which doesn't have a negative counterpart
+  def nonNegativeInt(rng: RNG): (Int, RNG) = {
+    val (i, r) = rng.nextInt
+    if (i < 0) (-i + 1, r) else (i, r)
+  }
+
+  // 6.2
+  // Generate a double between 0 and 1, not including 1.  Use Int.MaxValue and _.toDouble
+  def double(rng: RNG): (Double, RNG) = {
+    val (i, r) = nonNegativeInt(rng)
+    (i / (Int.MaxValue.toDouble + 1), r)
+  }
+
+  // 6.3 use functions already written
+  def intDouble(rng: RNG): ((Int,Double), RNG) = {
+    val (i, r) = rng.nextInt
+    val (d, r2) = double(r)
+    ((i, d), r2)
+  }
+
+  // 6.3 use functions already written
+  def doubleInt(rng: RNG): ((Double,Int), RNG) = {
+    val ((i, d), r) = intDouble(rng)
+    ((d, i), r)
+  }
+
+  // 6.3 use functions already written
+  def double3(rng: RNG): ((Double,Double,Double), RNG) = {
+    val (d1, r1) = double(rng)
+    val (d2, r2) = double(r1)
+    val (d3, r3) = double(r2)
+    ((d1, d2, d3), r3)
+  }
+
+  // 6.4 list of random Ints
+  def ints(count: Int)(rng: RNG): (List[Int], RNG) = {
+    if (count <= 0) (Nil, rng)
+    else {
+      val (i, r1) = rng.nextInt
+      val (l, r2) = ints(count - 1)(r1)
```

Comment:
It's not that important since it's just an exercise, but this wouldn't be stack safe. If the count is higher than the number of Stack Frame allowed in the runtime environment, this would throw an exception.